src/Formats/FormatFactory.cpp

```diff
@@ -166,6 +166,9 @@ BlockInputStreamPtr FormatFactory::getInput(
     // (segmentator + two parsers + reader).
     bool parallel_parsing = settings.input_format_parallel_parsing && file_segmentation_engine && settings.max_threads >= 4;
 
+    if (settings.min_chunk_bytes_for_parallel_parsing * settings.max_threads * 2 > settings.max_memory_usage)
```

Comment:
> BTW can't this be done by lowering the number of threads dynamically instead?

I don't know( I think there will be a complex logic because of corner cases. Chosen number of threads must be > 4, otherwise it will be ineffective. and if settings.max_memory_usage < settings.min_chunk_bytes_for_parallel_parsing, we have to turn it off...